litt/cli/push.go

```diff
@@ -0,0 +1,366 @@
+		knownHosts,
+		deleteAfterTransfer,
+		true,
+		threads,
+		throttleMB,
+		verbose)
+}
+
+// push uses rsync to transfer LittDB data to the remote location(s)
+func push(
+	logger logging.Logger,
+	sources []string,
+	destinations []string,
+	user string,
+	host string,
+	port uint64,
+	keyPath string,
+	knownHosts string,
+	deleteAfterTransfer bool,
+	fsync bool,
+	threads uint64,
+	throttleMB float64,
+	verbose bool) error {
+
+	if len(sources) == 0 {
+		return fmt.Errorf("no source paths provided")
+	}
+	if len(destinations) == 0 {
+		return fmt.Errorf("no destination paths provided")
+	}
+	if threads == 0 {
+		return fmt.Errorf("threads must be greater than 0")
+	}
+
+	// split bandwidth between workers
+	throttleMB /= float64(threads)
+
+	// Lock source files. It would be nice to also lock the remote directories, but that's tricky given that
+	// we are interacting with the remote machine via SSH and rsync.
+	releaseSourceLocks, err := util.LockDirectories(logger, sources, util.LockfileName, fsync)
+	if err != nil {
+		return fmt.Errorf("failed to lock source directories: %w", err)
+	}
+	defer releaseSourceLocks()
+
+	// Create an SSH session to the remote host.
+	connection, err := util.NewSSHSession(logger, user, host, port, keyPath, knownHosts, verbose)
+	if err != nil {
+		return fmt.Errorf("failed to create SSH session to %s@%s port %d: %w", user, host, port, err)
+	}
+
+	tables, err := lsPaths(logger, sources, false, fsync)
+	if err != nil {
+		return fmt.Errorf("failed to list tables in source paths %v: %w", sources, err)
+	}
+
+	for _, tableName := range tables {
+		err = pushTable(
+			logger,
+			tableName,
+			sources,
+			destinations,
+			connection,
+			deleteAfterTransfer,
+			fsync,
+			throttleMB,
+			threads,
+		)
+
+		if err != nil {
+			return fmt.Errorf("failed to push table %s: %w", tableName, err)
+		}
+	}
+
+	return nil
+}
+
+// Figure out which files are already present at the destination(s). Although these files may be partial, we always
+// want to preserve any pre-existing arrangements of files at the destination(s).
+//
+// The returned map is a map from file name (e.g. 1234.metadata) to the destination path (e.g. /path/to/remote/dir).
+func mapExistingFiles(
+	logger logging.Logger,
+	destinations []string,
+	tableName string,
+	connection *util.SSHSession) (map[string]string, error) {
+
+	existingFiles := make(map[string]string)
+
+	extensions := []string{segment.MetadataFileExtension, segment.KeyFileExtension, segment.ValuesFileExtension}
+
+	for _, dest := range destinations {
+		tableDestination := path.Join(dest, tableName, segment.SegmentDirectory)
+		filePaths, err := connection.FindFiles(tableDestination, extensions)
+		if err != nil {
+			return nil, fmt.Errorf("failed to list files in destination %s: %w", dest, err)
+		}
+
+		for _, filePath := range filePaths {
+			// Extract the file name from the path.
+			fileName := path.Base(filePath)
+			if _, exists := existingFiles[fileName]; !exists {
```

Comment:
Is this logic backwards? Looks like you're adding to `existingFiles` all files that _don't_ exist

Reply:
I think this is backwards. I didn't test this case, which explains how the bug slipped by. I'll add a test for this.

fixed

Reply:
Ok, so it turns out I did test this specifically, and those tests started failing when I inverted this condition. The variable naming is definitely bad here, since it confused even me for a while.

I think it should be more clear with the reframing:

```
			_, alreadyFoundAtDestination := existingFiles[fileName]
			enforce.False(alreadyFoundAtDestination,
				"duplicate file found: %s and %s", fileName, existingFiles[fileName])
			existingFiles[fileName] = dest
```

The variable formerly called `exists` (now `alreadyFoundAtDestination`) is only true if you find a file at the destination more than once. If it's not, it's probably a signal that there is data corruption, or will likely be data corruption if we continue.